pkg/util/retry/retry.go

```diff
@@ -116,53 +145,138 @@ func (r *Retry) Next() bool {
 		return true
 	}
 
-	if r.opts.MaxRetries > 0 && r.currentAttempt >= r.opts.MaxRetries {
+	backoff := r.retryIn()
+
+	if r.retryLimitReached() || r.shouldPreemptivelyCancel(backoff) {
 		return false
 	}
 
-	// Wait before retry.
-	d := r.retryIn()
-	if d > 0 {
-		log.VEventfDepth(r.ctx, 1 /* depth */, 2 /* level */, "will retry after %s", d)
+	backoffCh, maxDurExpCh, cleanup := r.timerCh(backoff)
+	defer cleanup()
+
+	log.VEventfDepth(
+		r.ctx, 1 /* depth */, 2 /* level */, "will retry after %s", backoff,
+	)
+
+	if r.backingOffHook != nil {
+		r.backingOffHook(backoff)
 	}
+
 	select {
-	case <-time.After(d):
-		r.currentAttempt++
-		return true
 	case <-r.opts.Closer:
 		return false
 	case <-r.ctx.Done():
 		return false
+	case <-backoffCh:
+		r.currentAttempt++
+		return true
+	case <-maxDurExpCh:
+		return false
 	}
 }
 
-// closedC is returned from Retry.NextCh whenever a retry
-// can begin immediately.
-var closedC = func() chan time.Time {
-	c := make(chan time.Time)
+func (r *Retry) retryLimitReached() bool {
+	return (r.opts.MaxRetries > 0 && r.currentAttempt >= r.opts.MaxRetries) ||
+		(r.opts.MaxDuration > 0 && !r.opts.Clock.Now().Before(r.deadline))
+}
+
+// immediateCh creates a channel that is immediately written to with the
+// provided value and then closed.
+func immediateCh(v bool) chan bool {
+	c := make(chan bool, 1)
+	c <- v
 	close(c)
 	return c
-}()
+}
+
+// timerCh returns two channels: one for the backoff timer and one for the
+// maximum duration expiration timer. It also returns a cleanup function that
+// stops both timers. A time.Time value is sent on the channels when their
+// respective timers expire.
+func (r *Retry) timerCh(
+	backoff time.Duration,
+) (backoffCh, maxDurExpCh <-chan time.Time, cleanup func()) {
```

Comment:
Not for this PR, but I kinda wonder if we could avoid creating new timers every time and instead always just have two timers on our retry struct that we reset.

Reply:
After implementing the other changes, just ended up throwing this change in here too because I couldn't help myself.